Sources/Gate/PolicyCollection.swift

```diff
@@ -0,0 +1,25 @@
+struct PolicyCollection<Ability: AbilitySet> {
+    var policies: [TypeTuple:Any] = [:]
```

Comment:
I'm not sure how many policies you expect to be storing, but a dictionary has a curve of performance that becomes more effective than an array once it contains "quite a lot" of entities. I'm not sure what the effective falloff is, and it would differ per type is used as a value in the Array or Dictionary. But it's always worth noting even if you don't apply it here. It might be useful knowledge in the future 😄 

I tried very hard, but I don't think you can work around `Any` here. This looks really good 👍

Reply:
Largely this is true, except that dictionary is the fastest way to access a value by key. Dictionary isn't any different, it also has to abide by the rules you just described like looping over values.

A computer is really a machine that knows integers and arrays. So a dictionary is a clever construction with arrays and integers. It, too, doesn't escape the limitations of your computer.

Dictionary is also called a `HashMap` because of it's optimization:

1. Make a hash based on the key (which is Hashable)
2. Find the first index in your array of hashes that matches this exact hash
3. Look at the keys associated with this hash and compare them to be exactly identical. Hashes are just a quick way of looking up but don't guarantee there are no duplicates
4. If the key matches, return the value associated with this key

An array is a bit simpler and uses less pointers (references) and this saves the overhead of the HashMap (Dictionary). There is a point where your hashmap becomes faster than an array, but because of how modern x86 CPUs work an array can be optimized better than a hashmap.